crypto/ube/fork_detect.c

```diff
@@ -37,13 +36,33 @@ OPENSSL_STATIC_ASSERT(MADV_WIPEONFORK == 18, MADV_WIPEONFORK_is_not_18)
 #define MADV_WIPEONFORK 18
 #endif
 
-DEFINE_STATIC_ONCE(g_fork_detect_once)
-DEFINE_STATIC_MUTEX(g_fork_detect_lock)
-DEFINE_BSS_GET(volatile char *, g_fork_detect_addr)
-DEFINE_BSS_GET(uint64_t, g_fork_generation)
-DEFINE_BSS_GET(int, g_ignore_madv_wipeonfork)
+static CRYPTO_once_t fork_detect_once = CRYPTO_ONCE_INIT;
+static struct CRYPTO_STATIC_MUTEX fork_detect_lock = CRYPTO_STATIC_MUTEX_INIT;
 
-static int init_fork_detect_madv_wipeonfork(void *addr, long page_size) {
+// This pointer is |volatile| because the value pointed to may be changed by
+// external forces (i.e. the kernel wiping the page) thus the compiler must not
+// assume that it has exclusive access to it.
+static volatile char *fork_detect_addr = NULL;
```

Comment:
Any reason why we model it as a `char *` here for the address but `uint32_t *` down in the snapsafe code?

Reply:
That's following the specified APIs.

The fork detection mechanisms is using the mapped page as a signal. The page has a special attribute enabled and will be zeroised if a fork event occurs. A char is sufficient. The generation number is maintained by us.

the snapsafe mechanism is used the stored data of the mapped page as the generation number. The value stored is maintained by the kernel and only a 32-bit number is exposed.